cranelift/filetests/src/zkasm_codegen.rs

```diff
@@ -281,68 +283,158 @@ pub fn compile_clif_function(func: &Function) -> Vec<String> {
+        arg_types.push_str(" ");
+
+        args_pushing.push_str(&format!("{arg_type}.const {arg}\n        "));
+    }
+    if arg_types.len() > 0 {
+        arg_types.pop();
+    }
+    // TODO: remove line with 8 whitespaces in the end of args_pushing
+    let wat_code = format!(
+        r#"(module
+    (import "env" "assert_eq" (func $assert_eq (param {res_bitness} {res_bitness})))
+    (import "env" "{func_name}" (func ${func_name} (param {arg_types}) (result {res_bitness})))
+    (func $main
+        {args_pushing}
+        call ${func_name}
+        {res_bitness}.const {expected_result}
+        call $assert_eq
+    )
+    (start $main)
+)"#,
+        args_pushing = args_pushing,
+        arg_types = arg_types,
+        res_bitness = res_bitness,
+        func_name = func_name,
+        expected_result = expected_result,
+    );
+    wat_code.to_string()
+}
+
 /// Compiles a invocation.
 pub fn compile_invocation(
     invoke: Invocation,
     compare: Comparison,
     expected: Vec<DataValue>,
 ) -> Vec<String> {
-    // Here I assume that each "function" in zkasm gets it's arguments from first N registers
-    // and put result in A.
-    // TODO: should be more robust way to do it, we need somehow define inputs and outputs
-    let mut res: Vec<String> = Default::default();
-    let registers = vec!["A", "B", "C", "D", "E"];
+    // TODO: don't do this clones
+    let cmp = if compare == Comparison::Equals {
+        Comparison::Equals
+    } else {
+        Comparison::NotEquals
+    };
+    let inv = Invocation {
+        func: invoke.func.clone(),
+        args: invoke.args.clone(),
+    };
+    let wat = runcommand_to_wasm(inv, cmp, expected.clone());
+    let wasm_module = wat::parse_str(wat).unwrap();
+
+    let settings = ZkasmSettings::default();
 
-    let args = invoke.args;
+    // TODO: we should not use generate_zkasm itself, but a bit changed version?
+    let generated: Vec<String> = generate_zkasm(&settings, &wasm_module)
+        .split("\n")
+        .map(|s| s.to_string())
+        .collect();
+    let new_label = invoke_name(&invoke);
     let funcname = invoke.func;
 
-    // TODO: here we should pay attention to type of DataValue (I64 or I32)
-    for (idx, arg) in args.iter().enumerate() {
-        res.push(format!("  {} => {}", arg, registers[idx]))
-    }
-    res.push(format!("    :JMP({})", funcname));
-    // TODO: handle functions with multiple outputs
-    res.push(format!("  {} => B", expected[0]));
-    // TODO: replace with call to host function
-    res.push(format!("  CALL AWESOME ASSERT ({})", compare));
-    res
+    // TODO: will this always function_2?
+    let start_index = generated.iter().position(|r| r == "function_2:").unwrap();
+    let end_index = generated.iter().rposition(|r| r == "  :JMP(RR)").unwrap();
+    let mut generated_function = generated[start_index..=end_index].to_vec();
+
+    generated_function[0] = format!("{}:", new_label);
+    // TODO: will this always function_1?
```

Comment:
What is `function_1` expected to be?

Reply:
When compiling such `.wat` snippets, result was look like (at least for programs I was testing):
```zkasm
;; preamble
function_2:
  ;; put values to the registers
  ;; JMP(function_1)
  ;; assert_helper
  ;; JMP(RR)
```

From one side, while `.wat` code have only have 3 functions, and all of them always appear in same order, it is quite natural that the functions always named same way. From the other side, I hope to find more reliable way to do it once finishing preamble generation.

Reply:
I thinks it's quite hard to follow what happens in `compile_invocation`, even more since it relies on implementation details of codegen. The explanations from your comments would be suited for comments in the code. Ideally the comments answer these questions:

- What is the invocation compiled to? To me it's not clear from context or return type.
- How is the generated zkASM expected to look like?
- What are `function_1` and `function_2` expected to be?

Reply:
This function generally seems like its trying to tip toe around our own code, which is a clear indication of an abstraction that's working against us. For instance, we appear to be parsing the output in an attempt to determine function boundaries, but nothing stops e.g. `JMP(RR)` to be used for something else than returning to a caller... LLVM for example especially likes a pattern similar to
```
    push rr
    rr <= addrof(done)
    call some_scope
some_scope:
    do_whatever_with_non-trivial control flow...
    jmp(rr)...

done:
    pop(rr)
```

within a single function in certain scenarios (e.g. to implement jumptables) and there is no reason why clift couldn't learn these tricks either.

IIRC cranelift already works on a per-function basis anyway so it might very well be worthwhile to revisit whether `generate_zkasm` makes sense if it interferes with use-cases as straightforward as this one. In fact from what I can see working around it is more code than replicating what `generate_zkasm` does in the first place... So perhaps splitting it up slightly would be a simpler approach?